### ISSUE TITLE
Error on svn commit "GitLab: LFS objects are missing. Ensure LFS is properly set up or try a manual"

### ISSUE THREAD

**Issue by DJm00n** [bozaro/git-as-svn#175]
I have an issue with commiting LFS tracked files to our GitLab CE server via git-to-svn.
SVN client have an error on commit:
```
Command: Commit  
Adding: D:\project\xxxxx\doc\temp2.uasset  
: D:\project\xxxxx\doc\temp2.uasset  
Committing transaction...:   
Error: Commit failed (details follow):  
Error: svn: E165001: Hook  
Error:  /var/opt/gitlab/git-data/repositories/root/xxxxx.git/hooks/pre-receive  
Error:  failed with output:  
Error: GitLab: LFS objects are missing. Ensure LFS is properly set up or try a manual  
Error:  "git lfs push --all".  
Error: If you want to break the lock, use the 'Check For Modifications' dialog or the repository browser.  

```
git-as-svn.log
```
2018-07-03T14:32:05,953  INFO [SvnServer-thread-2] (SvnServer.java:169) - New connection from: /172.16.11.36:22121
2018-07-03T14:32:06,312  INFO [SvnServer-thread-2] (SvnServer.java:319) - User: xxxx
2018-07-03T14:32:06,492  INFO [SvnServer-thread-2] (GitDeltaConsumer.java:178) - Created blob d919cdd810a0544c8258e1c7bb8c7558a41d5013 for file: /doc/temp2.uasset
2018-07-03T14:32:06,499  INFO [SvnServer-thread-2] (GitWriter.java:257) - Create commit 25d27f4ff17b3a2b9fd291759207defcbfbce7ed: test
2018-07-03T14:32:06,500  INFO [SvnServer-thread-2] (GitWriter.java:265) - Validate properties
2018-07-03T14:32:06,501  INFO [SvnServer-thread-2] (GitWriter.java:268) - Try to push commit in branch: refs/heads/develop
2018-07-03T14:32:06,614  WARN [SvnServer-thread-2] (CommitCmd.java:625) - Found error in cmd close-edit
org.tmatesoft.svn.core.SVNException: svn: E165001: Hook /var/opt/gitlab/git-data/repositories/root/xxxx.git/hooks/pre-receive failed with output:
GitLab: LFS objects are missing. Ensure LFS is properly set up or try a manual "git lfs push --all".

        at svnserver.repository.git.push.GitPushEmbedded.runHook(GitPushEmbedded.java:123) ~[git-as-svn.jar:?]
        at svnserver.repository.git.push.GitPushEmbedded.runReceiveHook(GitPushEmbedded.java:78) ~[git-as-svn.jar:?]
        at svnserver.repository.git.push.GitPushEmbedded.push(GitPushEmbedded.java:61) ~[git-as-svn.jar:?]
        at svnserver.repository.git.GitWriter$GitCommitBuilder.commit(GitWriter.java:269) ~[git-as-svn.jar:?]
        at svnserver.repository.git.GitWriter$GitCommitBuilder.commit(GitWriter.java:95) ~[git-as-svn.jar:?]
        at svnserver.server.command.CommitCmd$EditorPipeline.lambda$closeEdit$8(CommitCmd.java:563) ~[git-as-svn.jar:?]
        at svnserver.repository.locks.PersistentLockFactory.wrapLock(PersistentLockFactory.java:65) ~[git-as-svn.jar:?]
        at svnserver.repository.locks.PersistentLockFactory.wrapLockWrite(PersistentLockFactory.java:56) ~[git-as-svn.jar:?]
        at svnserver.repository.git.GitRepository.wrapLockWrite(GitRepository.java:687) ~[git-as-svn.jar:?]
        at svnserver.server.command.CommitCmd$EditorPipeline.closeEdit(CommitCmd.java:557) ~[git-as-svn.jar:?]
        at svnserver.server.command.LambdaCmd.process(LambdaCmd.java:46) ~[git-as-svn.jar:?]
        at svnserver.server.command.CommitCmd$EditorPipeline.editorCommand(CommitCmd.java:622) ~[git-as-svn.jar:?]
        at svnserver.server.SvnServer.serveClient(SvnServer.java:206) ~[git-as-svn.jar:?]
        at svnserver.server.SvnServer.lambda$run$0(SvnServer.java:172) ~[git-as-svn.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_171]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_171]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_171]
2018-07-03T14:32:06,632  WARN [SvnServer-thread-2] (SvnServer.java:226) - Command execution error: svn: E165001: Hook /var/opt/gitlab/git-data/repositories/root/xxxx.git/hooks/pre-receive failed with output:
GitLab: LFS objects are missing. Ensure LFS is properly set up or try a manual "git lfs push --all".
```
git-as-svn.conf almost same as https://github.com/bozaro/git-as-svn/blob/master/cfg/config-gitlab.example:
```
!config:
realm: Example realm
compressionEnabled: true
cacheConfig: !persistentCache
  path: /var/cache/git-as-svn/git-as-svn.mapdb
# Use GitLab repositories
repositoryMapping: !gitlabMapping
  path: /var/opt/gitlab/git-data/repositories/
  template:
    branch: develop
    renameDetection: true

# Use GitLab user database
userDB: !gitlabUsers {}

shared:
  # Web server settings
  # Used for:
  #  * detecticting add/remove repositories via GitLab System Hook
  #  * git-lfs-authenticate script (optionaly)
  - !web
    baseUrl: http://git.xxxxx.local/
    listen:
    - !http
      host: localhost
      port: 8123
      # Use X-Forwarded-* headers
      forwarded: true
  # GitLab LFS server
  - !lfs
    # Secret token for git-lfs-authenticate script
    # token: secret
    path: /var/opt/gitlab/gitlab-rails/shared/lfs-objects
    saveMeta: false
    compress: false
    layout: GitLab
  # GitLab server
  - !gitlab
    url: http://localhost:81/
    hookUrl: http://localhost:8123/
    token: xxxxxxx

```
Not LFS-tracked files are comming just fine via svn.
git commit working fine with all content.
Git as svn version: 1.2.0, f382c77c0e1e9c52326016586173c29c6ec0bc8e
GitLab version 10.2.5.
What should I do to fix it?

**Comment by slonopotamus** [comment]
Are files you're trying to commit appear in LFS? So, is it only GitLab failing to find them or other git clients also cannot see them?

**Comment by DJm00n** [comment]
Just checked and seems like git-as-svn successfully copied file with needed sha256 to /var/opt/gitlab/gitlab-rails/shared/lfs-objects folder structure and GitLab just failing to find it.

**Comment by DJm00n** [comment]
Found related GitLab issue and hook code:
https://gitlab.com/gitlab-org/gitlab-ce/issues/39836
https://gitlab.com/gitlab-org/gitlab-ce/blob/master/lib/gitlab/checks/change_access.rb#L172
https://gitlab.com/gitlab-org/gitlab-ce/blob/master/lib/gitlab/checks/lfs_integrity.rb

**Comment by DJm00n** [comment]
...but that will break GitLab's lfs project `all_lfs_objects` database as files are not linked to project... and git user will not be able to download files via usual GitLab's LFS API  (code [here](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/controllers/projects/lfs_api_controller.rb#L44) and [here](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/controllers/projects/lfs_api_controller.rb#L59)) like in #130.
So git-as-svn should call somehow [LfsLinkService](https://gitlab.com/gitlab-org/gitlab-ce/blob/master/app/services/projects/lfs_pointers/lfs_link_service.rb ) etc to properly link lfs files to project.
Is there any way to force git-as-svn put files to gitlab via usual `git lfs push` like http API?

**Comment by slonopotamus** [comment]
git-as-svn LFS code was written back in gitlab-9 days when gitlab LFS solution was in a very immature state. We need to review it and possibly change some stuff.

**Comment by slonopotamus** [comment]
We cannot simply `git lfs push`:

1. We need to get lfs url somewhere
2. We need to get authentication token so lfs http api accepts our push

**Comment by slonopotamus** [comment]
Alternatively, we could try calling LfsLinkService via gitlab-shell, though we also do not currently know where gitlab-shell is located.

**Comment by slonopotamus** [comment]
I believe this is fixed in git-as-svn 1.8.0 that was just released.

**Comment by slonopotamus** [comment]
Docs are lagging behind a bit, sorry.

Actually, this is all that is needed to make git-as-svn use GitLab's LFS server: https://github.com/bozaro/git-as-svn/blob/master/cfg/config-gitlab.example#L32
